src/BoundingBoxCameraSensor.cc

```diff
@@ -0,0 +1,525 @@
+bool BoundingBoxCameraSensor::Update(
+  const std::chrono::steady_clock::duration &_now)
+{
+  IGN_PROFILE("BoundingBoxCameraSensor::Update");
+  if (!this->dataPtr->initialized)
+  {
+    ignerr << "Not initialized, update ignored.\n";
+    return false;
+  }
+
+  if (!this->dataPtr->boundingboxCamera || !this->dataPtr->rgbCamera)
+  {
+    ignerr << "Camera doesn't exist.\n";
+    return false;
+  }
+
+  // don't render if there is no subscribers
+  if (!this->dataPtr->imagePublisher.HasConnections() &&
+    !this->dataPtr->boxesPublisher.HasConnections())
+  {
+    return false;
+  }
+
+  // Render the bounding box camera
+  this->Render();
+
+  // Publish image only if there is subscribers for it
+  if (this->dataPtr->imagePublisher.HasConnections())
+  {
+    // The sensor updates only the bounding box camera with its pose
+    // as it has the same name, so make rgb camera with the same pose
+    this->dataPtr->rgbCamera->SetWorldPose(
+      this->dataPtr->boundingboxCamera->WorldPose());
+
+    // Render the rgb camera
+    this->dataPtr->rgbCamera->Capture(this->dataPtr->image);
+
+    // Draw bounding boxes
+    for (auto box : this->dataPtr->boundingBoxes)
+    {
+      this->dataPtr->imageBuffer = this->dataPtr->image.Data<unsigned char>();
+
+      this->dataPtr->boundingboxCamera->DrawBoundingBox(
+        this->dataPtr->imageBuffer, box);
+    }
+
+    auto width = this->dataPtr->rgbCamera->ImageWidth();
+    auto height = this->dataPtr->rgbCamera->ImageHeight();
+
+    // Create Image message
+    this->dataPtr->imageMsg.set_width(width);
+    this->dataPtr->imageMsg.set_height(height);
+    // Format
+    this->dataPtr->imageMsg.set_step(
+      width * rendering::PixelUtil::BytesPerPixel(rendering::PF_R8G8B8));
+    this->dataPtr->imageMsg.set_pixel_format_type(
+      msgs::PixelFormatType::RGB_INT8);
+    // Time stamp
+    auto stamp = this->dataPtr->imageMsg.mutable_header()->mutable_stamp();
+    *stamp = msgs::Convert(_now);
+    auto frame = this->dataPtr->imageMsg.mutable_header()->add_data();
+    frame->set_key("frame_id");
+    frame->add_value(this->Name());
+    // Image data
+    this->dataPtr->imageMsg.set_data(this->dataPtr->imageBuffer,
+        rendering::PixelUtil::MemorySize(rendering::PF_R8G8B8,
+        width, height));
+
+    // Publish
+    this->AddSequence(this->dataPtr->imageMsg.mutable_header(), "rgbImage");
+    this->dataPtr->imagePublisher.Publish(this->dataPtr->imageMsg);
+  }
+
+  if (this->dataPtr->type == rendering::BoundingBoxType::Box3D)
+  {
+    this->dataPtr->boxes3DMsg.Clear();
+
+    // Create 3D boxes message
+    for (const auto &box : this->dataPtr->boundingBoxes)
+    {
+      // box data
+      auto annotated_box = this->dataPtr->boxes3DMsg.add_annotated_box();
+
+      auto oriented3DBox = new msgs::Oriented3DBox();
+      auto center = new msgs::Vector3d();
+      auto size = new msgs::Vector3d();
+      auto rotation = new msgs::Quaternion();
```

Comment:
Is there a chance for memory leaks here? It looks like these variables are added to the message, but when/how are they de-allocated?

Reply:
I don't know yet a way about how to deallocate it
is it deleted when we call the MSG_NAME.clear() function ?

I mean the whole msg, the AnnotatedAxisAligned2DBox_V